src/UrqlClient.re

```diff
@@ -0,0 +1,373 @@
+type t;
+
+/* Helpers for supporting polymorphic fetchOptions. */
+type fetchOptions('a) =
+  | FetchOpts(Fetch.requestInit): fetchOptions(Fetch.requestInit)
+  | FetchFn(unit => Fetch.requestInit)
+    : fetchOptions(unit => Fetch.requestInit);
+
+let unwrapFetchOptions =
+    (type a, fetchOptions: option(fetchOptions(a))): option(a) =>
+  switch (fetchOptions) {
+  | Some(FetchOpts(opts)) => Some(opts)
+  | Some(FetchFn(fn)) => Some(fn)
+  | None => None
+  };
+
+/* Helpers for supporting a user-supplied fetch function. */
+type fetchImpl('a) =
+  | FetchWithUrl(string => Js.Promise.t(Fetch.response))
+    : fetchImpl(string => Js.Promise.t(Fetch.response))
+  | FetchWithUrlOptions(
+      (string, Fetch.requestInit) => Js.Promise.t(Fetch.response),
+    )
+    : fetchImpl((string, Fetch.requestInit) => Js.Promise.t(Fetch.response))
+  | FetchWithRequest(Fetch.request => Js.Promise.t(Fetch.response))
+    : fetchImpl(Fetch.request => Js.Promise.t(Fetch.response))
+  | FetchWithRequestOptions(
+      (Fetch.request, Fetch.requestInit) => Js.Promise.t(Fetch.response),
+    )
+    : fetchImpl(
+        (Fetch.request, Fetch.requestInit) => Js.Promise.t(Fetch.response),
+      );
+
+let unwrapFetchImpl = (type a, fetch: option(fetchImpl(a))): option(a) => {
+  switch (fetch) {
+  | Some(FetchWithUrl(impl)) => Some(impl)
+  | Some(FetchWithUrlOptions(impl)) => Some(impl)
+  | Some(FetchWithRequest(impl)) => Some(impl)
+  | Some(FetchWithRequestOptions(impl)) => Some(impl)
+  | None => None
+  };
+};
+
+module Exchanges = {
```

Comment:
The `Exchanges` module is now a submodule of `Client` so each module can safely reference types from the other w/out a cyclic dependency error.